File: sensor_transports/transport_manager.py
#!/usr/bin/env python

import glob
import os
import sys
import random
import time
import numpy as np

from sensor_transports.camera_transporter import CameraTransporter
from sensor_transports.gnss_transporter import GnssTransporter
from sensor_transports.imu_transporter import ImuTransporter
from sensor_transports.lidar_transporter import LidarTransporter
from sensor_transports.radar_transporter import RadarTransporter

class TransportManager:
    def __init__(self, sensor_list):
        self.transporter_list = self.launch_sensor_transporter(sensor_list)
    def run_loop(self):
        pass

    def launch_sensor_transporter(self, sensor_list):
        transporter_list = []
        for sensor in sensor_list:
            if sensor['status'] == 'disabled':
                continue
            if sensor['type'] == 'camera':
                transporter_list.append(CameraTransporter(sensor))
            elif sensor['type'] == 'gnss':
                transporter_list.append(GnssTransporter(sensor))
            elif sensor['type'] == 'imu':
                transporter_list.append(ImuTransporter(sensor))
            elif sensor['type'] == 'lidar':
                transporter_list.append(LidarTransporter(sensor))
            elif sensor['type'] == 'radar':
                transporter_list.append(RadarTransporter(sensor))
        return transporter_list

File: sensor_transports/TransportManager.cpp
#include "TransportManager.hpp"
#include "CameraTransporter.hpp"
#include "GnssTransporter.hpp"
#include "ImuTransporter.hpp"
#include "LidarTransporter.hpp"
#include "RadarTransporter.hpp"

namespace sensor_transports
{
namespace transporter
{

TransportManager::TransportManager(const json& sensors_array)
{
    launchSensorTransporter(sensors_array);
}

TransportManager::~TransportManager()
{
    m_transporter_list.clear();
}

void TransportManager::runLoop()
{
    size_t cycle_count = 0;

    while (true)
    {
        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
        cycle_count++;
        std::cout << "TransportManager cycle_count: " << cycle_count << std::endl;
    }
}

void TransportManager::launchSensorTransporter(const json& sensors_array)
{
    for (const auto& sensor : sensors_array)
    {
        std::string sensor_type = sensor["type"];

        if (sensor_type.compare("camera") == 0)
        {
            std::cout << "launchCameraTransporter camera" << std::endl;
            m_transporter_list.push_back(std::make_unique<CameraTransporter>(sensor));
        }
        else if (sensor_type.compare("gnss") == 0)
        {
            m_transporter_list.push_back(std::make_unique<GnssTransporter>(sensor));
        }
        else if (sensor_type.compare("imu") == 0)
        {
            m_transporter_list.push_back(std::make_unique<ImuTransporter>(sensor));
        }
        else if (sensor_type.compare("lidar") == 0)
        {
            m_transporter_list.push_back(std::make_unique<LidarTransporter>(sensor));
        }
        else if (sensor_type.compare("radar") == 0)
        {
            m_transporter_list.push_back(std::make_unique<RadarTransporter>(sensor));
        }
    }
}
} // namespace transporter
} // namespace sensor_transports
